### PR TITLE
convert el.c to elem.d

### PR DESCRIPTION
**Pull request by WalterBright**

### CODE REVIEW

File: src/dmd/backend/elem.d [hunk 1]
```diff
@@ -444,13 +533,17 @@ void el_copy(elem *to,elem *from)
 
 elem * el_alloctmp(tym_t ty)
 {
-  symbol *s;
+    version (MARS)
+    { }
+    else
+        assert(!PARSER);
```
Comment (wilzbach) [inline src/dmd/backend/elem.d]: The formatting you used later: `version (MARS) { } else assert(!PARSER);` looks more natural.

File: src/dmd/backend/elem.d [hunk 2]
```diff
@@ -781,146 +897,149 @@ int el_anydef(elem *ed, elem *e)
...
+extern (C) // necessary because D <=> C++ mangling of "long long" is not consistent across memory models
 {
 elem * el_long(tym_t t,targ_llong val)
 { elem *e;
 
-  assert(MARS || !PARSER);
+    version (MARS)
+    { }
+    else
+        assert(!PARSER);
```
Comment (wilzbach) [inline src/dmd/backend/elem.d]: The formatting you used before `version (MARS) { } else assert(!PARSER);` looks more natural ;-)

File: src/dmd/backend/elem.d [hunk 3]
```diff
@@ -1667,41 +1814,52 @@ elem * el_var(symbol *s)
...
-    e->EV.sp.Vsym = s;
-    if (SCPP && PARSER)
-    {   type *t = s->Stype;
+    e.Eoper = OPvar;
+    e.EV.Vsym = s;
+
+    version (SCPP_HTOD)
+        enum scpp = true;
+    else
+        enum scpp = false;
```
Comment (wilzbach) [inline src/dmd/backend/elem.d]: Could have simplified a few if clauses before too.

File: src/dmd/backend/elem.d [hunk 4]
```diff
@@ -3076,46 +3282,81 @@ int el_allbits(elem *e,int bit)
...
 }
 
 /******************************
  * Extract long double value from constant elem.
  * Silently ignore types which are not floating point values.
  */
 
+version (CRuntime_Microsoft)
+{
+longdouble_soft el_toldouble(elem *e)
```
Comment (wilzbach) [inline src/dmd/backend/elem.d]:
```
..\dmd\backend\el.d(174): error : Function type does not match previously declared function with the same mangled name: `?el_toldouble@@YA?AUlongdouble_soft@@PAUelem@@@Z` [C:\projects\dmd\src\vcbuild\dmd.vcxproj]
```

The declaration is:

```
targ_ldouble el_toldouble(elem *);
```

(and not `longdouble_soft`)

File: src/dmd/backend/elem.d [hunk 5]
```diff
@@ -3328,63 +3581,68 @@ void elem_print_const(elem *e)
...
-#else
-            dbg_printf("%Lg ", e->EV.Vldouble);
-#endif
+version (CRuntime_Microsoft)
+{
+            char[3 + 3 * (targ_ldouble).sizeof + 1] buffer = void;
+            static if (is(typeof(e.EV.Vldouble) == real))
+                ld_sprint(buffer.ptr, 'g', longdouble_soft(e.EV.Vldouble));
+            else
+                ld_sprint(buffer.ptr, 'g', longdouble_soft(cast(real)e.EV.Vldouble));
```
Comment (wilzbach) [inline src/dmd/backend/elem.d]: Wouldn't the cast to `real` be irrelevant for the first case (i.e. if it's a real)?

File: src/dmd/backend/code.d [hunk 6]
```diff
@@ -178,6 +178,10 @@ struct linnum_data
 
 extern __gshared seg_data **SegData;
 
+ref targ_size_t Offset(int seg) { return SegData[seg].SDoffset; }
+ref targ_size_t Doffset() { return Offset(DATA); }
+ref targ_size_t CDoffset() { return Offset(CDATA); }
```
Comment (wilzbach) [inline src/dmd/backend/code.d]: This isn't used anywhere yet.

**Review by wilzbach — APPROVED** [review]
The AppVeyor failure is related to `longdouble_soft el_toldouble`.

File: src/dmd/backend/code.d (hunk 6 above)
Comment (WalterBright) [reply src/dmd/backend/code.d]: It will be.

**Pull request merged**